feat: show service banner in SSH/TTY sessions

code-asher opened the pull request:
- [X] Show banner
- [X] Test banner
- [x] Handle non-enterprise (from what I can tell our pattern is to just check the 404?)
- [x] Test auth changes
- [ ] Parse markdown?

Closes #7652.

code-asher committed:
Allow workspace agents to get appearance

code-asher committed:
Show service banner before MOTD

mafredri commented on agent/agentssh/agentssh.go:
```diff
@@ -63,9 +64,10 @@ type Server struct {
+	Env              map[string]string
+	AgentToken       func() string
+	Manifest         *atomic.Pointer[agentsdk.Manifest]
+	GetServiceBanner func(ctx context.Context) (codersdk.ServiceBannerConfig, error)
```
Suggestion: Drop the "get" to match other methods (e.g. `AgentToken`), see: https://golang.org/doc/effective_go.html#Getters
 
```suggestion
	ServiceBanner func(ctx context.Context) (codersdk.ServiceBannerConfig, error)
```

code-asher replied:
Ahhh very good to know, will change and thank you for the informative link!

mafredri commented on agent/agentssh/agentssh.go:
```diff
@@ -428,6 +435,19 @@ func (s *Server) startPTYSession(session ptySession, magicTypeLabel string, cmd
 }
 
+func (s *Server) showServiceBanner(ctx context.Context, session io.Writer) error {
+	banner, err := s.GetServiceBanner(ctx)
```
Consider checking `banner.Enabled` first to avoid a round-trip.

Since this will also slow down the start of every interactive SSH session by a server round-trip, should we try to push the change from the server and store it in the agent or poll it asynchronously? One option that comes to mind is piggy-backing on agent metadata.

code-asher replied:
This sounds very reasonable to me!  It would love if we can avoid the round trip.  I will look into metadata.

mafredri commented on agent/agentssh/agentssh.go:
```diff
@@ -428,6 +435,19 @@ func (s *Server) startPTYSession(session ptySession, magicTypeLabel string, cmd
+	if banner.Enabled && banner.Message != "" {
+		// The banner supports Markdown so we might want to parse it but Markdown is
+		// still fairly readable in its raw form.
+		_, err = io.WriteString(session, banner.Message+"\n\n\r")
```
Consider handling the error immediately to avoid a chance of code sneaking in between the if branch and `return err`.

code-asher replied:
Oh wow yeah that is way smarter.

mafredri commented on enterprise/coderd/appearance_test.go:
```diff
@@ -11,68 +11,125 @@ import (
+		banner, err = agentClient.GetServiceBanner(ctx)
+		require.NoError(t, err)
+		require.Equal(t, codersdk.ServiceBannerConfig{}, banner)
+	})
```
Since we have two optionals, should we add a test for when both are missing and it's expected behavior?

code-asher replied:
By the two optionals do you mean the status code checks?  I wanted to check the 404 but I was not sure how to run the test with the enterprise routes disabled.  Although now that I look at it a bit more, I think `coderdenttest` is for enterprise and `coderdtest` is non-enterprise?  I will try that.

Or do you mean a test for when the banner is missing and there is no license?

code-asher replied:
Turns out this is going away as I move the banner to be sent along with the other metadata anyway!

code-asher replied:
Er, nevermind, it is not going away just moving to another metadata endpoint.

mafredri reviewed (commented):
Nice feature and test coverage!

I would approve the change, but I think we should try to find a way to avoid the slow-down in initializing SSH sessions (see comment).

code-asher committed:
Drop non-idiomatic Get prefix

code-asher committed:
Handle error immediately

Otherwise there is a chance of code sneaking in between.

code-asher committed:
Test banner when enterprise endpoints are disabled

code-asher committed:
Poll for service banner

We need to provide the banner in the test so I changed the manifest
argument to support passing in the client.

We could not set the banner func afterward like I was doing before
because by that point the banner will already have fetched and will not
update until the next interval.

The test interval func also changes since I wanted to have a 15 second
interval but that meant 1.5 seconds in the test which is way too long.

code-asher commented:
Rebased to resolve a conflict but only the new commits are affected; the original two are still the same.

The banner is now polled (15 second interval at the moment).  I think pushing is better but the polling seemed easier.  I am open to pushing instead, just need to figure out the best way to do that.  Maybe just a new endpoint and coderd iterates over workspace agents and hits that endpoint on each to update the banner?

mafredri commented on agent/agentssh/agentssh.go:
```diff
@@ -763,6 +771,20 @@ func isQuietLogin(rawCommand string) bool {
+	if banner.Enabled && banner.Message != "" {
+		// The banner supports Markdown so we might want to parse it but Markdown is
+		// still fairly readable in its raw form.
+		_, err := io.WriteString(session, banner.Message+"\n\n\r")
```
I think we should inspect the newline count at the end of `banner.Message` and at most add two, WDYT? I'm worried we may add too much space because the banner was created with newlines at the end and that could look jarring (note that WebUI allows any number of newlines at the end).

PS. I think we also shouldn't output `\r` here. The SSH session (pty) should (hopefully) already be doing automatic `\n` -> `\r\n` translation if it was requested by the client.

At least on Mac, when I update the service banner in the WebUI, all newlines are simply `\n`, but it'd be possible to post any format via the API. So perhaps we should replace all `\r\n` with `\n` just in case, but we should verify this works as expected on Windows. Otherwise we might want to do dos2unix/unix2dos conversion depending on the platform (sigh).

mafredri replied:
Oh right, the pty isn't active yet at this point. So I misspoke about the `\r\n` translation. Looks like we'll have to replace all `\n` with `\r\n` even on Unix. We do the same with `showMOTD`.

I think you could use a `bufio.NewScanner(strings.Reader(banner.Message))` for convenience, or perhaps a two-way replace (`\n` => `\r\n` and `\r\r\n` => `\r\n`).

code-asher replied:
Good catch!  I completely missed the banner could already have newlines in it.

code-asher replied:
For the extra newline issue I went with a trim before adding the two newlines.

mafredri commented on agent/agentssh/agentssh.go:
```diff
@@ -346,6 +348,12 @@ func (s *Server) startPTYSession(session ptySession, magicTypeLabel string, cmd
 	session.DisablePTYEmulation()
 
 	if !isQuietLogin(session.RawCommand()) {
```
I think we should relax the constraint here since a service banner seems more important than a MOTD. Perhaps the presence of `.hushlogin` shouldn't silence it. The login shell check should be kept though.

code-asher replied:
Agreed!

mafredri commented on coderd/httpmw/workspaceagent.go:
```diff
@@ -18,40 +18,67 @@ import (
-	user, ok := r.Context().Value(workspaceAgentContextKey{}).(database.WorkspaceAgent)
+	user, ok := WorkspaceAgentOptional(r)
 	if !ok {
 		panic("developer error: agent middleware not provided")
```
This panic could now be confusing, since it's potentially provided but made optional.

mafredri commented on agent/agent.go:
```diff
@@ -491,6 +499,25 @@ func (a *agent) setLifecycle(ctx context.Context, state codersdk.WorkspaceAgentL
 	}
 }
 
+// fetchServiceBannerLoop fetches the service banner on an interval.
```
Perhaps we should add a note here that this won't immediately fetch anything and that it's expected to be primed elsewhere? (This way the intent is clear.)

code-asher replied:
I like this idea.

mafredri commented on agent/agent.go:
```diff
@@ -491,6 +499,25 @@ func (a *agent) setLifecycle(ctx context.Context, state codersdk.WorkspaceAgentL
 
+// fetchServiceBannerLoop fetches the service banner on an interval.
+func (a *agent) fetchServiceBannerLoop(ctx context.Context) {
+	ticker := time.NewTicker(adjustIntervalForTests(15*time.Second, time.Millisecond*100))
```
Should we increase this to like 2 minutes? 15 seconds seems awfully frequent for something that's not likely to change very often, and if it does there should be quite a bit of heads up (e.g. doing maintenance tomorrow).

code-asher replied:
Really good point.  I think I was overly concerned with someone starting a session right after the banner was updated but like you said there should be plenty of heads up.

mafredri reviewed (commented):
I think polling is fine for now! Sorry I sent you on the wrong path with the metadata/manifest. I was under the impression we updated it periodically but it seems we usually only do it once at startup.

A few more comments but I don't think any major changes are necessary here 👍🏻

code-asher commented on agent/agent.go:
```diff
@@ -491,6 +499,25 @@ func (a *agent) setLifecycle(ctx context.Context, state codersdk.WorkspaceAgentL
+		case <-ticker.C:
+			serviceBanner, err := a.client.GetServiceBanner(ctx)
+			if err != nil {
+				a.logger.Warn(ctx, "failed to fetch service banner")
```
I like the terminology change.  I did actually have this as `Error` before but it was failing a test.  From what I recall it seemed the agent was being shut down as the request was being made.  

Wait that probably just means I need to check `ctx.Err()` before logging the error.  I bet it was closed while the request was being made or something like that.

code-asher commented on agent/agentssh/agentssh.go:
```diff
@@ -346,6 +348,12 @@ func (s *Server) startPTYSession(session ptySession, magicTypeLabel string, cmd
 
 	if !isQuietLogin(session.RawCommand()) {
+		serviceBanner := s.ServiceBanner.Load()
+		err := showServiceBanner(session, serviceBanner)
```
Oops, good point, I had that at one point but apparently I lost it as I kept switching between passing in a pointer and a function.

code-asher committed:
Log banner update as an error

code-asher committed:
Check if service banner is nil

I managed to lose this in a scuffle.

code-asher committed:
Ensure carriage returns at the end of each line

Also trim to avoid adding too many newlines.

code-asher committed:
Reword agent panic

The panic might also be because you made the agent optional.

code-asher committed:
Note that the banner must be primed elsewhere

code-asher committed:
Increase banner fetch interval

code-asher commented:
> Sorry I sent you on the wrong path with the metadata/manifest.

Not at all!  It was still extremely useful since I was basically able to copy what we did for the manifest plus an extra polling loop.  <3

code-asher committed:
Make banner independent of .hushlogin

code-asher commented:
Not entirely sure what is going on with the Datadog update step but the tests seem to be passing, at least.

mafredri reviewed (approved):
Nice work!

code-asher committed:
fixup! Make banner independent of .hushlogin

matifali commented:
Awesome ♥️
